### PR TITLE
日々の進捗に応じてstreakなどを表示する

### PR DESCRIPTION
**Pull request by highwide**
streakなどを表示させるにあたり、「週ごとのHashのArray」っていう構造はつらかったので、RangeProgressesというclassを作りました。
そこのメソッドとして、streakの計算を行うものなど生やしています。
- [x] テスト(~_~;)

↑書いた！(2015/9/6 10:02)

### CODE REVIEW

File: app/controllers/practices_controller.rb [hunk 1]
```diff
@@ -7,8 +7,9 @@ def index
 
   def show
     @practice = Practice.find(params[:id])
-    progresses = @practice.daily_progresses.in_this_year
-    @progresses = format_for_graph_of_a_year(progresses)
+    @end_date = Time.zone.now.to_date
+    @start_date = @end_date.years_ago(1).to_date
+    @range_progresses = RangeProgresses.new(params[:id], @start_date, @end_date)
```
Comment (highwide) [inline app/controllers/practices_controller.rb:12]: **[reek]**
- [DuplicateMethodCall] calls params[:id] 2 times

File: app/helpers/practice_helper.rb [hunk 2]
```diff
@@ -1,8 +1,6 @@
 module PracticeHelper
   # TODO: 使ってみて緑の濃さ調整する
-  def color(daily)
-    return '#eee' if daily.blank?
-    progress_rate = daily.progress.to_f / daily.goal.to_f
+  def color(progress_rate)
```
Comment (highwide) [inline app/helpers/practice_helper.rb]: __[reek]__
* [UtilityFunction] doesn't depend on instance state

File: app/models/range_progresses.rb [hunk 3]
```diff
@@ -0,0 +1,68 @@
+class RangeProgresses
```
Comment (highwide) [inline app/models/range_progresses.rb:1]: **[reek]**
- [TooManyInstanceVariables] has at least 6 instance variables

File: app/models/range_progresses.rb [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+      start_date: longest[0],
+      end_date: longest[-1]
+    }
+  end
+
+  def current_streak
+    return @current if defined? @current
+
+    dates_from_current = @done_dates.reverse
+    if dates_from_current[0] == Time.zone.now.to_date
```
Comment (highwide) [inline app/models/range_progresses.rb]: **[reek]**
- [DuplicateMethodCall] calls dates_from_current[0] 4 times
- [FeatureEnvy] refers to dates_from_current more than self

File: app/models/range_progresses.rb [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+      streak = 1
+      start_date = dates_from_current[0]
```
Comment (highwide) [inline app/models/range_progresses.rb]: __[reek]__
* [DuplicateMethodCall] calls dates_from_current[0] 4 times
* [FeatureEnvy] refers to dates_from_current more than self

File: app/models/range_progresses.rb [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+      while dates_from_current[0] - 1.day == dates_from_current[1]
```
Comment (highwide) [inline app/models/range_progresses.rb:50]: **[reek]**
- [DuplicateMethodCall] calls dates_from_current[0] 4 times
- [FeatureEnvy] refers to dates_from_current more than self
- [FeatureEnvy] refers to dates_from_current more than self

File: app/helpers/practice_helper.rb (hunk 2 above)
Comment (highwide) [inline app/helpers/practice_helper.rb]: __[rails_best_practices]__
* remove unused methods (PracticeHelper#color)

File: app/models/range_progresses.rb [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+        dates_from_current = dates_from_current.drop(1)
+        start_date = dates_from_current[0]
```
Comment (highwide) [inline app/models/range_progresses.rb]: __[reek]__
* [DuplicateMethodCall] calls dates_from_current[0] 4 times
* [FeatureEnvy] refers to dates_from_current more than self

File: app/models/range_progresses.rb [hunk 8]
```diff
@@ -0,0 +1,68 @@
+class RangeProgresses
+  def initialize(id, start_date, end_date)
+    @start_date = start_date
+    @end_date = end_date
+    @progresses = DailyProgress.where(practice_id: id, done_at: start_date..end_date)
+    @done_dates = @progresses.where('progress > 0').pluck(:done_at).sort
+  end
+
+  def format_each_week
```
Comment (highwide) [inline app/models/range_progresses.rb]: __[rails_best_practices]__
* remove unused methods (RangeProgresses#format_each_week)

File: app/models/range_progresses.rb [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+      end
+    longest = dates_by_continuity.max_by{|dates| dates.count}
+    @longest = {
+      days: longest.count,
+      start_date: longest[0],
+      end_date: longest[-1]
+    }
+  end
+
+  def current_streak
```
Comment (highwide) [inline app/models/range_progresses.rb:42]: **[reek]**
- [TooManyStatements] has approx 9 statements

File: app/models/range_progresses.rb [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+    (@start_date..@end_date).each_with_object([]) do |date, formated_dates|
+      if formated_dates.blank? || formated_dates[-1].last.wday == 6
+        formated_dates << [date]
+      else
+        formated_dates[-1] << date
+      end
+    end
+  end
+
+  def total_done
```
Comment (highwide) [inline app/models/range_progresses.rb]: __[rails_best_practices]__
* remove unused methods (RangeProgresses#total_done)

File: app/models/range_progresses.rb [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+
+  def current_streak
+    return @current if defined? @current
+
+    dates_from_current = @done_dates.reverse
+    if dates_from_current[0] == Time.zone.now.to_date
+      streak = 1
+      start_date = dates_from_current[0]
+      while dates_from_current[0] - 1.day == dates_from_current[1]
+        dates_from_current = dates_from_current.drop(1)
```
Comment (highwide) [inline app/models/range_progresses.rb]: __[reek]__
* [FeatureEnvy] refers to dates_from_current more than self

File: app/models/range_progresses.rb [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+        formated_dates[-1] << date
+      end
+    end
+  end
+
+  def total_done
+    @done_dates.count
+  end
+
+  def longest_streak
```
Comment (highwide) [inline app/models/range_progresses.rb]: __[rails_best_practices]__
* remove unused methods (RangeProgresses#longest_streak)

File: app/models/range_progresses.rb [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+  def initialize(id, start_date, end_date)
+    @start_date = start_date
+    @end_date = end_date
+    @progresses = DailyProgress.where(practice_id: id, done_at: start_date..end_date)
+    @done_dates = @progresses.where('progress > 0').pluck(:done_at).sort
+  end
+
+  def format_each_week
+    (@start_date..@end_date).each_with_object([]) do |date, formated_dates|
+      if formated_dates.blank? || formated_dates[-1].last.wday == 6
```
Comment (highwide) [inline app/models/range_progresses.rb:11]: **[reek]**
- [DuplicateMethodCall] calls formated_dates[(-1)] 2 times
- [FeatureEnvy] refers to formated_dates more than self
- [FeatureEnvy] refers to formated_dates more than self

File: app/models/range_progresses.rb (hunk 9 above)
Comment (highwide) [inline app/models/range_progresses.rb]: __[rails_best_practices]__
* remove unused methods (RangeProgresses#current_streak)

File: app/models/range_progresses.rb [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+        formated_dates << [date]
+      else
+        formated_dates[-1] << date
```
Comment (highwide) [inline app/models/range_progresses.rb]: __[reek]__
* [DuplicateMethodCall] calls formated_dates[(-1)] 2 times
* [FeatureEnvy] refers to formated_dates more than self

File: app/models/range_progresses.rb [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+  end
+
+  def longest_streak
+    return @longest if defined? @longest
+
+    dates_by_continuity = \
+      @done_dates.each_with_object([]) do |date, dates_by_continuity|
+        if dates_by_continuity.blank? || dates_by_continuity.last.last + 1.day != date
+          dates_by_continuity << [date]
+        else
```
Comment (highwide) [inline app/models/range_progresses.rb]: **[rails_best_practices]**
- remove trailing whitespace [:ref](http://rails-bestpractices.com/posts/2010/12/02/remove-trailing-whitespace/)

File: app/models/range_progresses.rb [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+    @start_date = start_date
+    @end_date = end_date
+    @progresses = DailyProgress.where(practice_id: id, done_at: start_date..end_date)
+    @done_dates = @progresses.where('progress > 0').pluck(:done_at).sort
+  end
+
+  def format_each_week
+    (@start_date..@end_date).each_with_object([]) do |date, formated_dates|
+      if formated_dates.blank? || formated_dates[-1].last.wday == 6
+        formated_dates << [date]
```
Comment (highwide) [inline app/models/range_progresses.rb]: __[reek]__
* [FeatureEnvy] refers to formated_dates more than self

File: app/models/range_progresses.rb [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+  def total_done
+    @done_dates.count
+  end
+
+  def longest_streak
+    return @longest if defined? @longest
+
+    dates_by_continuity = \
+      @done_dates.each_with_object([]) do |date, dates_by_continuity|
+        if dates_by_continuity.blank? || dates_by_continuity.last.last + 1.day != date
```
Comment (highwide) [inline app/models/range_progresses.rb]: __[reek]__
* [DuplicateMethodCall] calls dates_by_continuity.last 2 times
* [FeatureEnvy] refers to dates_by_continuity more than self
* [FeatureEnvy] refers to dates_by_continuity more than self

File: app/models/range_progresses.rb [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+          dates_by_continuity << [date]
+        else 
+          dates_by_continuity.last << date
```
Comment (highwide) [inline app/models/range_progresses.rb]: __[reek]__
* [DuplicateMethodCall] calls dates_by_continuity.last 2 times
* [FeatureEnvy] refers to dates_by_continuity more than self

File: app/models/range_progresses.rb (hunk 12 above)
Comment (highwide) [inline app/models/range_progresses.rb:23]: **[reek]**
- [TooManyStatements] has approx 7 statements

File: app/models/range_progresses.rb [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+    @done_dates.count
+  end
+
+  def longest_streak
+    return @longest if defined? @longest
+
+    dates_by_continuity = \
+      @done_dates.each_with_object([]) do |date, dates_by_continuity|
+        if dates_by_continuity.blank? || dates_by_continuity.last.last + 1.day != date
+          dates_by_continuity << [date]
```
Comment (highwide) [inline app/models/range_progresses.rb]: __[reek]__
* [FeatureEnvy] refers to dates_by_continuity more than self

File: app/models/range_progresses.rb [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+        else 
+          dates_by_continuity.last << date
+        end
+      end
+    longest = dates_by_continuity.max_by{|dates| dates.count}
```
Comment (highwide) [inline app/models/range_progresses.rb]: __[reek]__
* [FeatureEnvy] refers to dates_by_continuity more than self

File: app/models/range_progresses.rb [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+      streak, start_date = 0, 'now!'
+    end
+    @current = {
+      days: streak,
+      start_date: start_date
+    }
+  end
+
+  def progress_rate(date)
+    daily = @progresses.to_a.select{|daily| daily.done_at == date}.first
```
Comment (highwide) [inline app/models/range_progresses.rb]: __[reek]__
* [FeatureEnvy] refers to daily more than self

File: app/models/range_progresses.rb [hunk 22]
```diff
@@ -0,0 +1,68 @@
...
+    return 0 if daily.blank?
```
Comment (highwide) [inline app/models/range_progresses.rb]: __[reek]__
* [FeatureEnvy] refers to daily more than self

File: app/models/range_progresses.rb [hunk 23]
```diff
@@ -0,0 +1,68 @@
...
+    daily.progress.to_f / daily.goal.to_f
```
Comment (highwide) [inline app/models/range_progresses.rb]: __[reek]__
* [FeatureEnvy] refers to daily more than self
* [FeatureEnvy] refers to daily more than self

File: app/helpers/practice_helper.rb (hunk 2 above)
Comment (highwide) [inline app/helpers/practice_helper.rb]: __[rails_best_practices]__
* remove unused methods (PracticeHelper#color)

Comment (highwide) [inline app/helpers/practice_helper.rb:3]: **[reek]**
- [UtilityFunction] doesn't depend on instance state

File: app/models/range_progresses.rb (hunk 8 above)
Comment (highwide) [inline app/models/range_progresses.rb:9]: **[rails_best_practices]**
- remove unused methods (RangeProgresses#format_each_week)

File: app/models/range_progresses.rb (hunk 10 above)
Comment (highwide) [inline app/models/range_progresses.rb:19]: **[rails_best_practices]**
- remove unused methods (RangeProgresses#total_done)

File: app/models/range_progresses.rb (hunk 12 above)
Comment (highwide) [inline app/models/range_progresses.rb:23]: **[rails_best_practices]**
- remove unused methods (RangeProgresses#longest_streak)

File: app/models/range_progresses.rb [hunk 24]
```diff
@@ -0,0 +1,68 @@
...
+  def current_streak
+    return @current if defined? @current
+
+    dates_from_current = @done_dates.reverse
+    if dates_from_current[0] == Time.zone.now.to_date
+      streak = 1
+      start_date = dates_from_current[0]
+      while dates_from_current[0] - 1.day == dates_from_current[1]
+        dates_from_current = dates_from_current.drop(1)
+        start_date = dates_from_current[0]
```
Comment (highwide) [inline app/models/range_progresses.rb:52]: **[reek]**
- [DuplicateMethodCall] calls dates_from_current[0] 4 times
- [FeatureEnvy] refers to dates_from_current more than self

File: app/models/range_progresses.rb [hunk 25]
```diff
@@ -0,0 +1,68 @@
...
+      end
+    longest = dates_by_continuity.max_by{|dates| dates.count}
+    @longest = {
+      days: longest.count,
+      start_date: longest[0],
+      end_date: longest[-1]
+    }
+  end
+
+  def current_streak
```
Comment (highwide) [inline app/models/range_progresses.rb:42]: **[rails_best_practices]**
- remove unused methods (RangeProgresses#current_streak)

File: app/models/range_progresses.rb [hunk 26]
```diff
@@ -0,0 +1,68 @@
...
+    return @current if defined? @current
+
+    dates_from_current = @done_dates.reverse
+    if dates_from_current[0] == Time.zone.now.to_date
+      streak = 1
+      start_date = dates_from_current[0]
+      while dates_from_current[0] - 1.day == dates_from_current[1]
+        dates_from_current = dates_from_current.drop(1)
```
Comment (highwide) [inline app/models/range_progresses.rb:51]: **[reek]**
- [FeatureEnvy] refers to dates_from_current more than self

File: app/models/range_progresses.rb (hunk 14 above)
Comment (highwide) [inline app/models/range_progresses.rb:14]: **[reek]**
- [DuplicateMethodCall] calls formated_dates[(-1)] 2 times
- [FeatureEnvy] refers to formated_dates more than self

File: app/models/range_progresses.rb (hunk 16 above)
Comment (highwide) [inline app/models/range_progresses.rb:12]: **[reek]**
- [FeatureEnvy] refers to formated_dates more than self

File: app/models/range_progresses.rb (hunk 17 above)
Comment (highwide) [inline app/models/range_progresses.rb:28]: **[reek]**
- [DuplicateMethodCall] calls dates_by_continuity.last 2 times
- [FeatureEnvy] refers to dates_by_continuity more than self
- [FeatureEnvy] refers to dates_by_continuity more than self

File: app/models/range_progresses.rb [hunk 27]
```diff
@@ -0,0 +1,68 @@
...
+
+  def longest_streak
+    return @longest if defined? @longest
+
+    dates_by_continuity = \
+      @done_dates.each_with_object([]) do |date, dates_by_continuity|
+        if dates_by_continuity.blank? || dates_by_continuity.last.last + 1.day != date
+          dates_by_continuity << [date]
+        else
+          dates_by_continuity.last << date
```
Comment (highwide) [inline app/models/range_progresses.rb:31]: **[reek]**
- [DuplicateMethodCall] calls dates_by_continuity.last 2 times
- [FeatureEnvy] refers to dates_by_continuity more than self

File: app/models/range_progresses.rb (hunk 19 above)
Comment (highwide) [inline app/models/range_progresses.rb:29]: **[reek]**
- [FeatureEnvy] refers to dates_by_continuity more than self

File: app/models/range_progresses.rb [hunk 28]
```diff
@@ -0,0 +1,68 @@
...
+        end
+      end
+    longest = dates_by_continuity.max_by{|dates| dates.count}
```
Comment (highwide) [inline app/models/range_progresses.rb:34]: **[reek]**
- [FeatureEnvy] refers to dates_by_continuity more than self

File: app/models/range_progresses.rb [hunk 29]
```diff
@@ -0,0 +1,68 @@
...
+      streak, start_date = 0, 'now!'
+    end
+    @current = {
+      days: streak,
+      start_date: start_date
+    }
+  end
+
+  def progress_rate(date)
+    daily = @progresses.to_a.select{|daily| daily.done_at == date}.first
```
Comment (highwide) [inline app/models/range_progresses.rb:65]: **[reek]**
- [FeatureEnvy] refers to daily more than self

File: app/models/range_progresses.rb [hunk 30]
```diff
@@ -0,0 +1,68 @@
...
+    return 0 if daily.blank?
```
Comment (highwide) [inline app/models/range_progresses.rb:66]: **[reek]**
- [FeatureEnvy] refers to daily more than self

File: app/models/range_progresses.rb [hunk 31]
```diff
@@ -0,0 +1,68 @@
...
+    daily.progress.to_f / daily.goal.to_f
```
Comment (highwide) [inline app/models/range_progresses.rb:67]: **[reek]**
- [FeatureEnvy] refers to daily more than self
- [FeatureEnvy] refers to daily more than self

File: app/models/range_progresses.rb (hunk 13 above)
Comment (highwide) [inline app/models/range_progresses.rb:11]: **[reek]**
- [FeatureEnvy] refers to formated_dates more than self
- [FeatureEnvy] refers to formated_dates more than self

File: app/models/range_progresses.rb (hunk 14 above)
Comment (highwide) [inline app/models/range_progresses.rb:14]: **[reek]**
- [FeatureEnvy] refers to formated_dates more than self

File: app/models/range_progresses.rb (hunk 17 above)
Comment (highwide) [inline app/models/range_progresses.rb:28]: **[reek]**
- [FeatureEnvy] refers to dates_by_continuity more than self
- [FeatureEnvy] refers to dates_by_continuity more than self

File: app/models/range_progresses.rb (hunk 27 above)
Comment (highwide) [inline app/models/range_progresses.rb:31]: **[reek]**
- [FeatureEnvy] refers to dates_by_continuity more than self

File: app/models/range_progresses.rb [hunk 32]
```diff
@@ -0,0 +1,69 @@
...
+      end_date: longest[-1]
+    }
+  end
+
+  def current_streak
+    return @current if defined? @current
+
+    dates_from_current = @done_dates.reverse
+    # FIXME: DuplicateMethodCall dates_from_current[0]
+    if dates_from_current[0] == Time.zone.today
```
Comment (highwide) [inline app/models/range_progresses.rb:47]: **[reek]**
- [DuplicateMethodCall] calls dates_from_current[0] 4 times
- [FeatureEnvy] refers to dates_from_current more than self

File: app/models/range_progresses.rb [hunk 33]
```diff
@@ -0,0 +1,69 @@
...
+      streak = 1
+      start_date = dates_from_current[0]
```
Comment (highwide) [inline app/models/range_progresses.rb:49]: **[reek]**
- [DuplicateMethodCall] calls dates_from_current[0] 4 times
- [FeatureEnvy] refers to dates_from_current more than self

File: app/models/range_progresses.rb [hunk 34]
```diff
@@ -0,0 +1,69 @@
...
+      while dates_from_current[0] - 1.day == dates_from_current[1]
```
Comment (highwide) [inline app/models/range_progresses.rb:50]: **[reek]**
- [DuplicateMethodCall] calls dates_from_current[0] 4 times
- [FeatureEnvy] refers to dates_from_current more than self
- [FeatureEnvy] refers to dates_from_current more than self

File: app/models/range_progresses.rb [hunk 35]
```diff
@@ -0,0 +1,69 @@
...
+        dates_from_current = dates_from_current.drop(1)
```
Comment (highwide) [inline app/models/range_progresses.rb:51]: **[reek]**
- [FeatureEnvy] refers to dates_from_current more than self

File: test/models/range_progresses.rb [hunk 36]
```diff
@@ -0,0 +1,53 @@
...
+    start_date = '2015-08-21'.to_date
+    end_date = Time.zone.today
+    progress = daily_progresses :start_of_range_progresses
+    @range_progresses = RangeProgresses.new(progress.practice.id, start_date, end_date)
+  end
+
+  # FIXME: 2016/8/22以降このテストはfailする
+  test '#format_each_week' do
+    result = @range_progresses.format_each_week
+    expectation1 = [
```
Comment (highwide) [inline test/models/range_progresses.rb]: __[reek]__
* [UncommunicativeVariableName] has the variable name 'expectation1'

File: test/models/range_progresses.rb [hunk 37]
```diff
@@ -0,0 +1,53 @@
...
+      '2015-08-21'.to_date, # 金曜
+      '2015-08-22'.to_date  # 土曜
+    ]
+    expectation2 = [
```
Comment (highwide) [inline test/models/range_progresses.rb]: __[reek]__
* [UncommunicativeVariableName] has the variable name 'expectation2'

File: test/models/range_progresses.rb [hunk 38]
```diff
@@ -0,0 +1,52 @@
...
+  def setup
+    start_date = '2015-08-28'.to_date
+    end_date = Time.zone.today
+    progress = daily_progresses :start_of_range_progresses
+    @range_progresses = RangeProgresses.new(progress.practice.id, start_date, end_date)
+  end
+
+  test '#format_each_week' do
+    result = @range_progresses.format_each_week
+    expectation1 = [
```
Comment (highwide) [inline test/models/range_progresses.rb]: __[reek]__
* [UncommunicativeVariableName] has the variable name 'expectation1'

File: test/models/range_progresses.rb [hunk 39]
```diff
@@ -0,0 +1,52 @@
...
+      '2015-08-28'.to_date, # 金曜
+      '2015-08-29'.to_date  # 土曜
+    ]
+    expectation2 = [
```
Comment (highwide) [inline test/models/range_progresses.rb]: __[reek]__
* [UncommunicativeVariableName] has the variable name 'expectation2'

File: test/models/range_progresses.rb (hunk 38 above)
Comment (highwide) [inline test/models/range_progresses.rb]: __[reek]__
* [UncommunicativeVariableName] has the variable name 'expectation1'

File: test/models/range_progresses.rb (hunk 39 above)
Comment (highwide) [inline test/models/range_progresses.rb]: __[reek]__
* [UncommunicativeVariableName] has the variable name 'expectation2'

File: app/controllers/practices_controller.rb (hunk 1 above)
Comment (ta1kt0me) [inline app/controllers/practices_controller.rb:12]: このあたり使うと短くできそう。
https://github.com/asakusarb/action_args#filters

File: app/helpers/practice_helper.rb (hunk 2 above)
Comment (ta1kt0me) [inline app/helpers/practice_helper.rb:3]: 無視して良いのでは。

File: app/models/range_progresses.rb (hunk 3 above)
Comment (ta1kt0me) [inline app/models/range_progresses.rb:1]: 無視して良いのでは。

File: app/models/range_progresses.rb (hunk 8 above)
Comment (ta1kt0me) [inline app/models/range_progresses.rb:9]: 同じく無視して良いでしょ

File: app/models/range_progresses.rb (hunk 10 above)
Comment (ta1kt0me) [inline app/models/range_progresses.rb:19]: 消しちゃいましょー

File: app/models/range_progresses.rb (hunk 12 above)
Comment (ta1kt0me) [inline app/models/range_progresses.rb:23]: 無視しましょう

File: app/controllers/practices_controller.rb (hunk 1 above)
Comment (highwide) [inline app/controllers/practices_controller.rb:12]: やべ使うくせなかった。ありがとうございます！

**Comment by ta1kt0me** [comment]
なんかコメントいろいろ無視して良いと思います。

**Pull request closed without merging**